codegen/winterfell/src/air/boundary_constraints.rs

```diff
@@ -55,125 +31,57 @@ pub(super) fn add_fn_get_aux_assertions(impl_ref: &mut Impl, ir: &AirIR) {
-                    .any(|input| input.0 == vector_access.name())
-                {
-                    format!("self.{}[{}]", vector_access.name(), vector_access.idx())
-                } else if is_aux_constraint {
-                    format!("E::from({}[{}])", vector_access.name(), vector_access.idx())
-                } else {
-                    format!("{}[{}]", vector_access.name(), vector_access.idx())
-                }
-            }
-            Self::MatrixAccess(matrix_access) => {
-                if is_aux_constraint {
-                    format!(
-                        "E::from({}[{}][{}])",
-                        matrix_access.name(),
-                        matrix_access.row_idx(),
-                        matrix_access.col_idx()
-                    )
-                } else {
-                    format!(
-                        "{}[{}][{}]",
-                        matrix_access.name(),
-                        matrix_access.row_idx(),
-                        matrix_access.col_idx()
-                    )
-                }
-            }
-            Self::Rand(index) => {
-                format!("aux_rand_elements.get_segment_elements(0)[{index}]")
-            }
-            Self::Add(lhs, rhs) => {
-                format!(
-                    "({}) + ({})",
-                    lhs.to_string(ir, is_aux_constraint),
-                    rhs.to_string(ir, is_aux_constraint)
-                )
-            }
-            Self::Sub(lhs, rhs) => {
-                format!(
-                    "({}) - ({})",
-                    lhs.to_string(ir, is_aux_constraint),
-                    rhs.to_string(ir, is_aux_constraint)
-                )
-            }
-            Self::Mul(lhs, rhs) => {
-                format!(
-                    "({}) * ({})",
-                    lhs.to_string(ir, is_aux_constraint),
-                    rhs.to_string(ir, is_aux_constraint)
-                )
-            }
-            Self::Exp(lhs, rhs) => {
-                if let Self::Const(rhs) = **rhs {
-                    format!("({}).exp({})", lhs.to_string(ir, is_aux_constraint), rhs)
-                } else {
-                    todo!()
-                }
-            }
-            _ => panic!("boundary constraint expressions cannot reference the trace"),
+    // add the boundary constraints
+    let mut last_row_declared = false;
+    for constraint in ir.boundary_constraints(trace_segment) {
+        let (trace_access, expr_root) = ir
+            .constraint_graph()
+            .split_boundary_constraint(constraint.node_index())
+            .expect("failed to split boundary constraint");
+        debug_assert!(trace_access.trace_segment() == trace_segment);
+
+        // declare the last row variable, if needed.
+        if !last_row_declared && constraint.domain() == ConstraintDomain::LastRow {
+            func_body.line(format!("let {last_row} = self.last_step();"));
+            last_row_declared = true;
```

Comment:
Thought for the future: it seems like determining if the `last_row` variable was declared or not adds quite a bit of complexity. Should we just always use `self.last_step()` instead? If it has `inline(always)` attribute, should theoretically give the same result.

Reply:
I'm happy to update this now. I actually originally declared `last_row` in all cases specifically to avoid this complexity, but this solution seems like the best option